sdk/monitor/azure-monitor-ingestion/src/main/java/com/azure/monitor/ingestion/LogsIngestionAsyncClient.java

```diff
@@ -0,0 +1,268 @@
+    public Mono<UploadLogsResult> upload(String dataCollectionRuleId, String streamName,
+                                         List<Object> logs, UploadLogsOptions options) {
+        return withContext(context -> upload(dataCollectionRuleId, streamName, logs, options, context));
+    }
+
+    Mono<UploadLogsResult> upload(String dataCollectionRuleId, String streamName,
+                                  List<Object> logs, UploadLogsOptions options,
+                                  Context context) {
+        return Mono.defer(() -> splitAndUpload(dataCollectionRuleId, streamName, logs, options, context));
+    }
+
+    private Mono<UploadLogsResult> splitAndUpload(String dataCollectionRuleId, String streamName, List<Object> logs, UploadLogsOptions options, Context context) {
+        try {
+            Objects.requireNonNull(dataCollectionRuleId, "'dataCollectionRuleId' cannot be null.");
+            Objects.requireNonNull(dataCollectionRuleId, "'streamName' cannot be null.");
+            Objects.requireNonNull(dataCollectionRuleId, "'logs' cannot be null.");
+
+            if (logs.isEmpty()) {
+                throw LOGGER.logExceptionAsError(new IllegalArgumentException("'logs' cannot be empty."));
+            }
+
+            ObjectSerializer serializer = DEFAULT_SERIALIZER;
+
+            // set concurrency to 1 as default
+            int concurrency = 1;
+            if (options != null) {
+                if (options.getObjectSerializer() != null) {
+                    serializer = options.getObjectSerializer();
+                }
+                if (options.getMaxConcurrency() != null) {
+                    concurrency = options.getMaxConcurrency();
+                }
+            }
+
+            List<List<Object>> logBatches = new ArrayList<>();
+            // TODO (srnagar): can improve memory usage by creating these request payloads right before sending the
+            //  request if the allowed concurrency is lower than the number of requests.
+            List<byte[]> requests = createGzipRequests(logs, serializer, logBatches);
+            RequestOptions requestOptions = new RequestOptions()
+                    .addHeader(CONTENT_ENCODING, GZIP)
+                    .setContext(context);
+
+            Iterator<List<Object>> logBatchesIterator = logBatches.iterator();
+            return Flux.fromIterable(requests)
+                    .flatMapSequential(bytes ->
+                            uploadToService(dataCollectionRuleId, streamName, requestOptions, bytes), concurrency)
+                    .map(responseHolder -> mapResult(logBatchesIterator, responseHolder))
+                    .collectList()
+                    .map(this::createResponse);
+        } catch (RuntimeException ex) {
+            return monoError(LOGGER, ex);
+        }
+    }
+
+    private UploadLogsResult mapResult(Iterator<List<Object>> logBatchesIterator, UploadLogsResponseHolder responseHolder) {
```

Comment:
Could the functionality of this method be merged into `uploadToService` by having `List<Object> logsBatch` being passed into `uploadToService`?